tests/platform_tests/test_auto_negotiation.py

```diff
@@ -0,0 +1,356 @@
+                pytest_assert(fanout_actual_speed == speed, 'expect fanout speed: {}, but got {}'.format(speed, fanout_actual_speed))
+
+
+def test_force_speed():
+    """Test all candidate ports to force to all supported speeds one by one and verify
+       that the port operational status is up after auto negotiation
+    """
+    for duthost, candidates in cadidate_test_ports.items():
+        if not candidates:
+            continue
+        logger.info('Test candidate ports are {}'.format(candidates))
+        for dut_port, fanout, fanout_port in candidates:
+            logger.info('Start test for DUT port {} and fanout port {}'.format(dut_port, fanout_port))
+            # Disable auto negotiation on fanout port
+            success = fanout.set_auto_negotiation_mode(fanout_port, False)
+            if not success:
+                # Fanout does not support set auto negotiation mode for this port
+                logger.info('Ignore port {} due to fanout port {} does not support setting auto-neg mode'.format(dut_port, fanout_port))
+                continue
+
+            logger.info('Trying to get a common supported speeds set among dut port, fanout port and cable')
+            supported_speeds = get_supported_speeds_for_port(duthost, dut_port, fanout, fanout_port)
+            if not supported_speeds:
+                continue
+
+            logger.info('Run test based on supported speeds: {}'.format(supported_speeds))
+            duthost.shell('config interface autoneg {} disabled'.format(dut_port))
+            for speed in supported_speeds:
+                success = fanout.set_speed(fanout_port, speed)
+                if not success:
+                    logger.info('Skip speed {} because fanout does not support it'.format(speed))
+                    continue
+                duthost.shell('config interface speed {} {}'.format(dut_port, speed))
+                logger.info('Wait until the port status is up, expected speed: {}'.format(speed))
+                wait_result = wait_until(SINGLE_PORT_WAIT_TIME, 
+                                        PORT_STATUS_CHECK_INTERVAL, 
+                                        check_ports_up, 
+                                        duthost, 
+                                        [dut_port],
+                                        speed)
+                pytest_assert(wait_result, '{} are still down'.format(dut_port))
+                fanout_actual_speed = fanout.get_speed(fanout_port)
+                pytest_assert(fanout_actual_speed == speed, 'expect fanout speed: {}, but got {}'.format(speed, fanout_actual_speed))
+
+
+def get_cable_supported_speeds_helper(duthost):
+    """Get a cable supported speeds helper
+
+    Args:
+        duthost: DUT object
+
+    Returns:
+        object: A helper class or instance
+    """
+    asic_type = duthost.facts["asic_type"]
+
+    if asic_type == "mellanox":
+        return MlnxCableSupportedSpeedsHelper
```

Comment:
Is this also support 3rd party cable installed on Mellanox switch?

Reply:
Yes, it is supported according to my test.